.github/workflows/publish-release.yml

```
@@ -0,0 +1,239 @@
+        uses: actions/checkout@v4
+        with:
+          ref: ${{needs.prepare-release.outputs.RELEASE_TAG}}
+
+      - name: Set up JDK 8
+        uses: actions/setup-java@v4
+        with:
+          java-version: '8'
+          distribution: 'corretto'
+
+      - uses: actions/setup-python@v5 
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - name: Cache Maven packages
+        id: cache-maven
+        uses: actions/cache@v4
+        with:
+          path: ~/.m2/repository
+          key: ${{ runner.os }}-mvn-build-${{ matrix.params.spark-version }}-${{ matrix.params.scala-version }}-${{ hashFiles('pom.xml') }}
+          restore-keys: ${{ runner.os }}-mvn-build-${{ matrix.params.spark-version }}-${{ matrix.params.scala-version }}-
+      
+      - name: Cache Pip packages
+        id: cache-pip
+        uses: actions/cache@v4
+        with:
+          path: ~/.cache/pip
+          key: ${{ runner.os }}-pip-whl-${{ env.PYTHON_VERSION }}-${{ matrix.params.spark-version }}
+
+      - name: Import GPG key
+        id: sign
+        uses: crazy-max/ghaction-import-gpg@01dd5d3ca463c7f10f7f4f7b4f177225ac661ee4
+        with:
+          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+
+      - name: Publish artifacts
+        id: publish
+        run: |
+
+          # build and publish artifact
+          ./set-version.sh ${{ matrix.params.spark-version }} ${{ matrix.params.scala-version }}
+          mvn clean deploy -Dsign
+
+          # For Scala version 2.12 build & release python
+          if [[ ${{ matrix.params.scala-version }}  == 2.12* ]]; then
+              echo "Scala version starts with '2.12'"
+              ./build-whl.sh
+
+              # upload to test PyPi
+              pip install twine
+              twine check python/dist/*
+              python3 -m twine upload --repository testpypi python/dist/*
+
+              # upload to PyPi
+              python3 -m twine upload python/dist/*
+          fi
+
+      - name: Extract release notes
+        id: release-notes
+        run: |
+          awk '
+          BEGIN { capture = 0; print_capture = 0; }
+          /^## / {
+            if (capture == 0) {
+              capture = 1;              # Start capture on the first occurrence of "##"
+            } else if (capture == 1) {
+              print_capture = 1;        # Prepare to stop capturing on the next "##"
+            }
+          }
+
+          capture && !print_capture { print; }
+          print_capture { exit; }       # Exit after capturing the desired content
+          ' CHANGELOG.md > ./release-notes.txt
+
+          # Grab release name
+          name=$(grep "^## " release-notes.txt | head -n 1 | sed "s/^## //")
+          echo "release_name=$name" >> $GITHUB_OUTPUT
+
+          # Leave only body for the release
+          sed -i -e '1d' ./release-notes.txt # Remove the first line in the file
+          echo "release_notes_path=release-notes.txt" >> $GITHUB_OUTPUT
```

I think this can be made a bit more concise:
```suggestion
          grep -A 1000 -m 1 "^## " CHANGELOG.md | tail -n +2 | grep -B 1000 -m 1 "^## " | head -n -1 > ./release-notes.txt

          # Grab release name
          name=$(grep -m 1 "^## " CHANGELOG.md | sed "s/^## //")
          echo "release_name=$name" >> $GITHUB_OUTPUT

          # provide release notes file path as output
          echo "release_notes_path=release-notes.txt" >> $GITHUB_OUTPUT
```

---

Not sure why this has to work on more than one OS but fine with me.